File: vuln-sandbox/node/cwe-90/ldap-injection-safe.js
// SAFE: ldap-injection — LDAP special characters escaped before query construction
// Rule: InjectLdap | CWE-90 | Expected: TrueNegative

const express = require('express');
const ldap = require('ldapjs');

const app = express();

function escapeLdapFilter(value) {
  // SAFE: escape all LDAP special characters per RFC 4515
  return value
    .replace(/\\/g, '\\5c')
    .replace(/\*/g, '\\2a')
    .replace(/\(/g, '\\28')
    .replace(/\)/g, '\\29')
    .replace(/\0/g, '\\00');
}

app.get('/user', (req, res) => {
  const username = req.query.username;
  const client = ldap.createClient({ url: 'ldap://localhost:389' });

  // SAFE: username escaped before being used in the LDAP filter
  const safeUsername = escapeLdapFilter(username);
  const filter = `(uid=${safeUsername})`;

  client.search('dc=example,dc=com', { filter, scope: 'sub' }, (err, result) => {
    const entries = [];
    result.on('searchEntry', (entry) => entries.push(entry.object));
    result.on('end', () => { client.destroy(); res.json(entries); });
  });
});

app.listen(3000);

File: vuln-sandbox/python/cwe-90/ldap-injection-safe.py
# SAFE: ldap-injection — LDAP special characters escaped before filter construction
# Rule: InjectLdap | CWE-90 | Expected: TrueNegative

import ldap
import re
from flask import Flask, request, jsonify

app = Flask(__name__)

LDAP_SPECIAL_CHARS = re.compile(r'[\\*\(\)\x00]')


def escape_ldap_filter(value: str) -> str:
    """Escape LDAP filter special characters per RFC 4515."""
    replacements = {'\\': '\\5c', '*': '\\2a', '(': '\\28', ')': '\\29', '\x00': '\\00'}
    return LDAP_SPECIAL_CHARS.sub(lambda m: replacements[m.group()], value)


@app.route('/user')
def get_user():
    username = request.args.get('username', '')

    conn = ldap.initialize('ldap://localhost:389')
    conn.simple_bind_s('cn=admin,dc=example,dc=com', 'admin_password')

    # SAFE: username escaped before being used in the LDAP filter
    safe_username = escape_ldap_filter(username)
    ldap_filter = f'(uid={safe_username})'

    results = conn.search_s('dc=example,dc=com', ldap.SCOPE_SUBTREE, ldap_filter)
    conn.unbind_s()

    return jsonify([{'dn': dn, 'attrs': attrs} for dn, attrs in results])


if __name__ == '__main__':
    app.run()
